modules/assim.batch/R/pda.emulator.R

```diff
@@ -12,22 +10,23 @@
 ##' @author Ryan Kelly, Istem Fer
 ##' @export
 pda.emulator <- function(settings, params.id=NULL, param.names=NULL, prior.id=NULL, chain=NULL, 
-                     iter=NULL, adapt=NULL, adj.min=NULL, ar.target=NULL, jvar=NULL, n.knot=NULL) {
+                     iter=NULL, adapt=NULL, adj.min=NULL, ar.target=NULL, jvar=NULL, n.knot=NULL, burnin=NULL) {
   
   ## this bit of code is useful for defining the variables passed to this function 
   ## if you are debugging
   if(FALSE){
     params.id <- param.names <- prior.id <- chain <- iter <- NULL 
-    n.knot <- adapt <- adj.min <- ar.target <- jvar <- NULL
+    n.knot <- adapt <- adj.min <- ar.target <- jvar <- burnin <- NULL
   }
 
   ## -------------------------------------- Setup ------------------------------------- ##
   ## Handle settings
     settings <- pda.settings(
                   settings=settings, params.id=params.id, param.names=param.names, 
                   prior.id=prior.id, chain=chain, iter=iter, adapt=adapt, 
-                  adj.min=adj.min, ar.target=ar.target, jvar=jvar, n.knot=n.knot)
+                  adj.min=adj.min, ar.target=ar.target, jvar=jvar, n.knot=n.knot, burnin=burnin)
 
+   burnin <- ifelse(!is.null(settings$assim.batch$burnin), settings$assim.batch$burnin, ceiling(min(2000,0.2*settings$assim.batch$iter)))
```

Comment:
What’s wrong with ifelse, looks clean to me and a direct descendent of the ?: operator in C or Java.

What I really want to ban from R is undefined non-local scoping - violates fundamental principles in theory of computer science*. In R, a variable can be defined in a level up (or in global) and called by a function without it explicitly being passed, and changing this value changes it everywhere. Makes it really really hard to catch spelling errors of similarly named variables or when debugging and defining many variables on the command line to test things. Also leads to lazy coding where some of our functions in the workflow just magically pull variables (like settings) without requesting them. It’s also not easily possible (as far as I know) to be able to choose whether to pass variables by reference or value (except by a couple hacks or R package). 

-ankur
* Based on the little I learned back in 1996 when I took two 2 classes on the topic where I learned to derive formal lambda calculus and Turing diagrams

> On Jul 28, 2016, at 9:20 AM, Ben Bond-Lamberty <notifications@github.com> wrote:
> 
> In modules/assim.batch/R/pda.emulator.R <https://github.com/PecanProject/pecan/pull/941#discussion_r72630812>:
> 
> >  
> > +   burnin <- ifelse(!is.null(settings$assim.batch$burnin), settings$assim.batch$burnin, ceiling(min(2000,0.2*settings$assim.batch$iter)))
> If I could remove/ban ifelse from R, I would.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub <https://github.com/PecanProject/pecan/pull/941/files/d19d432de0100bde83a778f012340d5fce5b7309#r72630812>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AC8PlwnPZqusv8XwmBaJB2O5C0l-V2rKks5qaLqsgaJpZM4JWz5S>.
>